### DISCUSSION TITLE
Can I import a self-generated python script on runtime?

### DISCUSSION THREAD

**Discussion by ToasterUwU** [pyinstaller/pyinstaller#6655] [Help] [answered]
My layout is like this:

main.py - A script that based on some things generates python code and puts it in a .py file
internal_code.py - The autogenerated code

What I want is to run the code within that internal_code.py.
It works if I just run main.py, but it doesn't if I use "pyinstaller -F main.py" to make it executable and then run it. The error is:
```
import internal_code
ModuleNotFoundError: No module named 'internal_code'
```
The same happens when I use `internal_code = importlib.import_module("internal_code")` instead.

Is what I'm trying possible? The closest question I found was if it is possible to run static python code by importing it, which works, at least with some extra stuff while using pyinstaller like "paths". But that is not an option if I generate my code on the fly.

And I know what some of you might think: "Why can't you just code logic into main.py that does the same as the code you generate? That should be even easier than doing it your way."
In this very specific case, I cant. Explaining why would take some time that would be wasted since I know dynamically making the code is my only option.

Any ideas and answers are appreciated.

**Comment by bwoodsend** [comment, marked as answer]
> In this very specific case, I cant. Explaining why would take some time that would be wasted since I know dynamically making the code is my only option.

I seen so many people come to this conclusion at some point and every time they've been wrong so I'm afraid that I don't believe you.

When is this code generated? Before you run PyInstaller or at runtime? PyInstaller won't stop you doing either but if it's at runtime then you'll have to be careful where you write that generated file to because there's a good chance that the rest of your code will be installed into a directory requiring privileged access to write to.

The rules for importing files are pretty straight forward and are identical for PyInstaller as they are for non-PyInstaller: If a module is in a directory and that directory is in `sys.path` then you can import it.

**Reply by ToasterUwU** [reply]
In my case, I'm playing around with obfuscation.

Specifically embedding code as some obfuscated string, as an example in Roman Numerals, and then decoding it, writing it to a file, and running it.

There is no other way to go about this. I want a string inside a base program that I fill with some things to tell it what to convert and how to convert, then run pyinstaller on it, run it and see it running that code that was not included as actual code.

But thanks for the hint to sys.path, I forgot about that. I added some code to append the internal_code.py to it and then the import works.

I totally forgot that the behavior might be different with pyinstaller and it maybe won't be able to see things next to the .exe file I run.

Thanks for giving my brain a little kick.

**Comment by RandySabatin** [comment]
you can use Pyinstaller and "importlib.import_module" to run your script at run time. I made a sample demo at https://github.com/RandySabatin/PyScriptRunner
